File: backend/app/services/fees.py
from decimal import Decimal, ROUND_HALF_UP
from sqlalchemy.orm import Session
from app.config import settings as app_settings
from app.models.store_settings import StoreSettings


def get_settings(db: Session) -> StoreSettings:
    """Return the singleton store settings, creating it (from config defaults)
    on first use."""
    row = db.query(StoreSettings).filter(StoreSettings.id == "default").first()
    if not row:
        row = StoreSettings(
            id="default",
            delivery_fee=Decimal(str(app_settings.DELIVERY_FEE)),
            free_delivery_threshold=Decimal(str(app_settings.FREE_DELIVERY_THRESHOLD)),
            small_cart_fee=Decimal(str(app_settings.SMALL_CART_FEE)),
            small_cart_threshold=Decimal(str(app_settings.SMALL_CART_THRESHOLD)),
            tax_percent=Decimal("5"),
        )
        db.add(row)
        db.commit()
        db.refresh(row)
    return row


def compute_delivery_fee(subtotal: Decimal, s: StoreSettings) -> Decimal:
    """Delivery + small-cart handling fee from the store settings."""
    fee = Decimal("0")
    if subtotal < s.free_delivery_threshold:
        fee += s.delivery_fee
    if subtotal < s.small_cart_threshold:
        fee += s.small_cart_fee
    return fee


def compute_tax(subtotal: Decimal, s: StoreSettings) -> Decimal:
    """Tax on the item subtotal, rounded to 2 decimals."""
    tax = (subtotal * s.tax_percent / Decimal("100"))
    return tax.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

File: frontend/src/config/fees.ts
import { StoreSettings } from '../types';

// Fallback used before the admin-configured settings load from the backend.
export const DEFAULT_STORE_SETTINGS: StoreSettings = {
  delivery_fee: 25,
  free_delivery_threshold: 199,
  small_cart_fee: 15,
  small_cart_threshold: 99,
  tax_percent: 5,
};

/** Delivery + small-cart handling fee for a subtotal, per store settings. */
export function computeDeliveryFee(subtotal: number, s: StoreSettings): number {
  let fee = 0;
  if (subtotal < s.free_delivery_threshold) fee += Number(s.delivery_fee);
  if (subtotal < s.small_cart_threshold) fee += Number(s.small_cart_fee);
  return fee;
}

/** Tax on the item subtotal (2-decimal), per store settings. */
export function computeTax(subtotal: number, s: StoreSettings): number {
  return Math.round(subtotal * Number(s.tax_percent)) / 100;
}

/** How much more to add to unlock free delivery (0 once reached). */
export function amountToFreeDelivery(subtotal: number, s: StoreSettings): number {
  return Math.max(0, Number(s.free_delivery_threshold) - subtotal);
}
